default/python/AI/ProductionAI.py

```diff
@@ -1812,3 +1724,204 @@ def get_number_of_existing_outpost_and_colony_ships() -> int:
+
+
+def _try_enqueue(
+    building_type: BuildingType, candidates: Iterable[PlanetId], *, at_front: bool = False, ignore_dislike: bool = False
+) -> float:
+    """
+    Enqueue building at one of the planets in candidates.
+    If at_front, building is added at the front of the queue.
+    Returns PP per turn spent on the new building or 0.0 if nothing was enqueued.
+    """
+    universe = fo.getUniverse()
+    empire = fo.getEmpire()
+    opinion = building_type.get_opinions()
+    locations = []
+    preferred_locations = []
+    for pid in candidates:
+        planet = universe.getPlanet(pid)
+        if not planet:
+            error(f"Got pid {pid} in candidate, which does not seem to be a planetID")
+            continue
+        if not ignore_dislike and pid in opinion.dislikes:
+            continue
+        if not building_type.can_be_produced(pid) or not building_type.can_be_enqueued(pid):
+            continue
+        if pid in opinion.likes:
+            preferred_locations.append((planet.currentMeterValue(fo.meterType.maxTroops), pid))
+        else:
+            locations.append((planet.currentMeterValue(fo.meterType.maxTroops), pid))
+    for _, pid in sorted(preferred_locations) + sorted(locations):
+        planet = universe.getPlanet(pid)
+        res = building_type.enqueue(pid)
+        debug("Enqueueing %s at planet %d (%s) , with result %d", building_type, pid, planet.name, res)
+        if res:
+            if at_front:
+                res = fo.issueRequeueProductionOrder(empire.productionQueue.size - 1, 0)  # move to front
+                debug("Requeueing %s to front of build queue, with result %d", building_type, res)
+            return building_type.turn_cost(pid)
+    return 0.0
+
+
+def _may_enqueue_for_stability(building_type: BuildingType, new_turn_cost: float) -> float:
+    """
+    Build building if it seems worth doing so to increase stability.
+    Only builds of locations.planets_enqueued is empty and new_turn_cost is 0.0,
+    i.e. there are currently no build queue entries for the given building.
+    returns new_turn_cost or turn_cost of the building enqueued by this function.
+    """
+    if building_type.queued_in() or new_turn_cost:
+        return new_turn_cost
+    # this can be improved a lot, taking into account value of planets, actual stability and
+    # what effects the change would have. For the moment, keep it simple.
+    # Note that the strongest effect is always on the building's planet itself.
+    opinion = building_type.get_opinions()
+    universe = fo.getUniverse()
+    if len(opinion.likes) >= len(opinion.dislikes):
+        like_candidates = opinion.likes - building_type.built_at(True)
+        # plans may change, so consider only actual colonies that like it
+        candidates = [pid for pid in like_candidates if universe.getPlanet(pid).speciesName]
+        return _try_enqueue(building_type, candidates)
+    return 0.0
+
+
+def _build_scanning_facility() -> float:
```

Comment:
Have you considered using the class approach here? 

```python
class BaseBuildingEnqueue(ABC):
     @abstractmethod
     def get_building_type() -> BuildingType: ...

     def avaliable(self): return self.get_building_type().available()  
     
     @abstractmethod
     def _process_buiding_type(self): ...
     
     def process(self) -> float:
         if not self.avaliable():
            return 0.0   
         return slelf._process()

```        

The `process` will do the work.  

It will have zero type-specific code, just call its methods.   There are two kinds of methods:  abstract, like `_process`.  It will contain code that is unique for all building types.  And with reasonable defaults, so for most classes, it will be the same, but you could still override it for special cases. 

We could merge this code as is, just to have some examples of what use-cases might look like, and convert this to class later.

Reply:
With the class approach, we actually could iterate all these objects twice.  The first one is for normal enqueue, and the second for the `_may_enqueue_for_stability`,  to utilize resources that are left from the normal process.

Reply:
Let's do that later. When I've done more buildings, it's easier to see what they all have in common.